### PR TITLE
Add complex dtypes support for `tf.squared_difference`

### LINKED ISSUE #14932: tf.squared_difference does not work with complex dtypes
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes, the provided minimal example
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary (tensorflow-gpu)
- **TensorFlow version (use command below)**: 1.4.0
- **Python version**: 3.6
- **CUDA/cuDNN version**: 8/6
- **GPU model and memory**: GeForce GTX 980 4GB

### Describe the problem

`tf.squared_difference` does not work with complex dtypes `tf.complex64` and `tf.complex128` although the documentation says so: https://www.tensorflow.org/api_docs/python/tf/squared_difference.

### Source code / logs

Minimal (not) working example:

```python
import tensorflow as tf

for dtype in [tf.complex128, tf.complex64]:
    try:
        print(f'*************** {dtype} ***************')
        x = tf.constant(1, dtype=dtype)
        y = tf.constant(2, dtype=dtype)
        result = tf.squared_difference(x, y)

        with tf.Session() as sess:
            sess.run(result)
    except Exception as e:
        print(e)
```

The error output:

```
*************** <dtype: 'complex128'> ***************
2017-11-28 10:33:08.695995: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX
2017-11-28 10:33:08.836822: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties: 
name: GeForce GTX 980 major: 5 minor: 2 memoryClockRate(GHz): 1.2785
pciBusID: 0000:04:00.0
totalMemory: 3.94GiB freeMemory: 3.86GiB
2017-11-28 10:33:08.836855: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 980, pci bus id: 0000:04:00.0, compute capability: 5.2)
No OpKernel was registered to support Op 'SquaredDifference' with these attrs.  Registered devices: [CPU,GPU], Registered kernels:
  device='GPU'; T in [DT_INT32]
  device='GPU'; T in [DT_INT64]
  device='GPU'; T in [DT_DOUBLE]
  device='GPU'; T in [DT_HALF]
  device='GPU'; T in [DT_FLOAT]
  device='CPU'; T in [DT_INT64]
  device='CPU'; T in [DT_INT32]
  device='CPU'; T in [DT_DOUBLE]
  device='CPU'; T in [DT_HALF]
  device='CPU'; T in [DT_FLOAT]

         [[Node: SquaredDifference = SquaredDifference[T=DT_COMPLEX128](Const, Const_1)]]

Caused by op 'SquaredDifference', defined at:
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/net/home/neumann/workspace/python/tensorflow/tf_speech/tf_speech/complex_squared_difference_example.py", line 9, in <module>
    res = tf.squared_difference(x, y)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/site-packages/tensorflow/python/ops/gen_math_ops.py", line 4601, in squared_difference
    "SquaredDifference", x=x, y=y, name=name)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 2956, in create_op
    op_def=op_def)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1470, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): No OpKernel was registered to support Op 'SquaredDifference' with these attrs.  Registered devices: [CPU,GPU], Registered kernels:
  device='GPU'; T in [DT_INT32]
  device='GPU'; T in [DT_INT64]
  device='GPU'; T in [DT_DOUBLE]
  device='GPU'; T in [DT_HALF]
  device='GPU'; T in [DT_FLOAT]
  device='CPU'; T in [DT_INT64]
  device='CPU'; T in [DT_INT32]
  device='CPU'; T in [DT_DOUBLE]
  device='CPU'; T in [DT_HALF]
  device='CPU'; T in [DT_FLOAT]

         [[Node: SquaredDifference = SquaredDifference[T=DT_COMPLEX128](Const, Const_1)]]

*************** <dtype: 'complex64'> ***************
2017-11-28 10:33:08.858858: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: GeForce GTX 980, pci bus id: 0000:04:00.0, compute capability: 5.2)
No OpKernel was registered to support Op 'SquaredDifference' with these attrs.  Registered devices: [CPU,GPU], Registered kernels:
  device='GPU'; T in [DT_INT32]
  device='GPU'; T in [DT_INT64]
  device='GPU'; T in [DT_DOUBLE]
  device='GPU'; T in [DT_HALF]
  device='GPU'; T in [DT_FLOAT]
  device='CPU'; T in [DT_INT64]
  device='CPU'; T in [DT_INT32]
  device='CPU'; T in [DT_DOUBLE]
  device='CPU'; T in [DT_HALF]
  device='CPU'; T in [DT_FLOAT]

         [[Node: SquaredDifference_1 = SquaredDifference[T=DT_COMPLEX64](Const_2, Const_3)]]

Caused by op 'SquaredDifference_1', defined at:
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/net/home/neumann/workspace/python/tensorflow/tf_speech/tf_speech/complex_squared_difference_example.py", line 9, in <module>
    res = tf.squared_difference(x, y)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/site-packages/tensorflow/python/ops/gen_math_ops.py", line 4601, in squared_difference
    "SquaredDifference", x=x, y=y, name=name)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 2956, in create_op
    op_def=op_def)
  File "/home/neumann/.conda/envs/project_dc/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1470, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

InvalidArgumentError (see above for traceback): No OpKernel was registered to support Op 'SquaredDifference' with these attrs.  Registered devices: [CPU,GPU], Registered kernels:
  device='GPU'; T in [DT_INT32]
  device='GPU'; T in [DT_INT64]
  device='GPU'; T in [DT_DOUBLE]
  device='GPU'; T in [DT_HALF]
  device='GPU'; T in [DT_FLOAT]
  device='CPU'; T in [DT_INT64]
  device='CPU'; T in [DT_INT32]
  device='CPU'; T in [DT_DOUBLE]
  device='CPU'; T in [DT_HALF]
  device='CPU'; T in [DT_FLOAT]

         [[Node: SquaredDifference_1 = SquaredDifference[T=DT_COMPLEX64](Const_2, Const_3)]]
```

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #14932 where complex dtypes are not supported for `tf.squared_difference`, which is different from the doc string in `math_ops.cc` (see `BINARY_FEWER`).

This fix adds the `complex64` and `complex128` support in kernel, and adds additional test cases.

This fix fixes #14932.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by drpngx** [comment]
@benoitsteiner WDYT?

**Comment by carlthome** [comment]
Does this actually work as it should? The tests all have zeroes for the imaginary part so it's hard to tell.

**Comment by yongtang** [comment]
@carlthome The PR has been updated with additional test cases to cover imag != 0 case.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @benoitsteiner: It has been 41 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @benoitsteiner: It has been 79 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @benoitsteiner: It has been 94 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by qlzh727** [comment]
Nagging @benoitsteiner again.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @benoitsteiner: It has been 14 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer%(reviewers): You have been added as a reviewer to this pull request. Please add your review or reassign. It has been  29ays with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @benoitsteiner: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @benoitsteiner: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by wt-huang** [comment]
@rmlarsen Could you take a look?

File: tensorflow/python/ops/math_ops_test.py [hunk 1]
```diff
@@ -203,14 +203,25 @@ class SquaredDifferenceTest(test_util.TensorFlowTestCase):
...
       with test_util.device(use_gpu=True):
         z_tf = self.evaluate(math_ops.squared_difference(x, y))
         self.assertAllClose(z, z_tf)
 
+  @test_util.run_in_graph_and_eager_modes()
+  def testComplexSquaredDifference(self):
+    for dtype in [np.complex64, np.complex128]:
+      x = np.array([[1+3j, 2+2j, 3+1j], [4-1j, 5-2j, 6-3j]], dtype=dtype)
+      y = np.array([-3+1j, -2+2j, -1+3j], dtype=dtype)
+      z = (x - y) * (x - y)
```
Comment (rmlarsen) [inline tensorflow/python/ops/math_ops_test.py]: normally when we talk about "squared difference" we are referring to the L2 metric in a vector space. Therefore, this function should compute conj(x-y)*(x-y). In other words, you need to modify the implementation as well by adding a conjugation to the first input for complex types.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

**Comment by yongtang** [comment]
Thanks @rmlarsen for the review. I updated the PR to capture the complex support for squared difference. Please take a look at the change and see if it fits.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @rmlarsen, @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yongtang** [comment]
@rmlarsen The PR has been updated with `conj(x-y)*(x-y)` used now. Please take a look and see if there are any issues.

**Comment by harshini-gadige** [comment]
@rmlarsen  Could you please review and approve this PR.

**Review by rmlarsen — APPROVED** [review]

**Pull request merged**